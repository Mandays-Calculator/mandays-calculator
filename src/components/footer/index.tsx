export { default as Footer } from "./Footer";
export { StyledFooterContainer, StyledFooterText } from "./styles";
