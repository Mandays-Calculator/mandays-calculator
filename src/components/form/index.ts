export * from "./checkbox";
export * from "./checkbox-group";
export * from "./text-field";
export * from "./date-picker";
