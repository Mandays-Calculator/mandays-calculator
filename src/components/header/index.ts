export {
  StyledAppBar,
  StyledToolBarContainer,
  StyledSearchSelect,
  popperStyle,
  StyledProjectTitle,
} from "./styles";
