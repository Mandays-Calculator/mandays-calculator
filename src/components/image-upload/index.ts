export { default as ImageUpload } from "./ImageUpload";
export * from "./styles";
