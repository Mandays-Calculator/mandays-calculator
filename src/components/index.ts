export * from "./form/text-field";
export * from "./form/Form";
export * from "./title";
