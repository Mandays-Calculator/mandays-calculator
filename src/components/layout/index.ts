export { default as Layout } from "./Layout";
export {
  StyledLayoutContainer,
  StyledLayoutContent,
  ContentContainer,
} from "./styles";
