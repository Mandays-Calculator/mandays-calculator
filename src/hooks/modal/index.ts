export * from "./useDisabledBodyScroll";
export * from "./useUnauthorizedModal";
