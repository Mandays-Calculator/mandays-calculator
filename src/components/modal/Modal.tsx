import type { ReactNode, Ref, ReactElement, SyntheticEvent } from "react";
import type { TransitionProps } from "@mui/material/transitions";
import type { DialogProps } from "@mui/material/Dialog";
import type { SxProps } from "@mui/material";

import { forwardRef } from "react";
import { useDisableBodyScroll } from "~/hooks/modal";

import styled from "@mui/system/styled";
import Stack from "@mui/material/Stack";
import Box from "@mui/material/Box";

import Dialog from "@mui/material/Dialog";
import MuiDialogTitle from "@mui/material/DialogTitle";
import MuiDialogContent from "@mui/material/DialogContent";
import Slide from "@mui/material/Slide";

interface ModalProps extends DialogProps {
  open: boolean;
  title: string;
  maxWidth: "xs" | "sm" | "md" | "lg";
  dividers?: boolean;
  actions?: ReactNode;
  children: ReactNode;
  onClose: () => void;
  onEnter?: () => void;
  onExited?: () => void;
  sx?: SxProps;
}

const Transition = forwardRef(
  (
    transitionProps: TransitionProps & { children: ReactElement },
    ref: Ref<unknown>
  ) => {
    const { in: slideIn, children } = transitionProps;
    return (
      <Slide
        direction="up"
        in={slideIn}
        ref={ref}
        children={children}
        tabIndex={-1}
      />
    );
  }
);

const DialogActions = styled(Box)(({ theme }) => ({
  padding: theme.spacing(0, 2, 2, 2),
}));

const DialogTitle = styled(MuiDialogTitle)(({ theme }) => ({
  display: "flex",
  alignItems: "center",
  justifyContent: "space-between",
  padding: theme.spacing(2, 0, 2, 2),
  color: theme.palette.primary.main,
}));

const DialogContent = styled(MuiDialogContent)(({ theme }) => ({
  padding: theme.spacing(2),
}));

const Offset = styled(Box)(({ theme }) => ({
  width: theme.spacing(2.5),
}));

type ReasonType = "backdropClick" | "escapeKeyDown";

export const Modal = (props: ModalProps): ReactElement => {
  const {
    open,
    title,
    dividers,
    actions,
    children,
    maxWidth = "sm",
    onClose,
    onEnter,
    onExited,
    sx,
  } = props;

  useDisableBodyScroll(open);

  const handleBackDrop = (_event: SyntheticEvent, reason: ReasonType): void => {
    if (reason !== "backdropClick") {
      onClose();
    }
  };

  return (
    <Dialog
      TransitionComponent={Transition}
      TransitionProps={{ onEnter, onExited }}
      sx={{ maxWidth: "100%", ...sx }}
      disableScrollLock={true}
      keepMounted={false}
      maxWidth={maxWidth}
      open={open}
      onClose={handleBackDrop}
    >
      <Stack direction="row" display="flex" minHeight={35}>
        <Box flexGrow={1}>
          {title ? <DialogTitle>{title}</DialogTitle> : null}
        </Box>
        <Offset />
      </Stack>
      <DialogContent dividers={dividers}>{children}</DialogContent>
      {actions ? <DialogActions>{actions}</DialogActions> : null}
    </Dialog>
  );
};

export default Modal;
