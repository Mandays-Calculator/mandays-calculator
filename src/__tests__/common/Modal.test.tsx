import { fireEvent, render } from "@testing-library/react";
import { Modal } from "~/components/modal";
import ProviderWrapper from "./utils/ProviderWrapper";
import Button from "@mui/material/Button";

jest.mock("~/hooks/modal", () => ({
  useDisableBodyScroll: jest.fn(),
}));

describe("Modal Component", () => {
  it("renders and matches snapshot", () => {
    const { asFragment } = render(
      <ProviderWrapper>
        <Modal open={true} title="Test Modal" onClose={() => {}} maxWidth="sm">
          <div>Modal content goes here</div>
        </Modal>
      </ProviderWrapper>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders actions when provided", () => {
    const onCloseMock = jest.fn();
    const actionText = "Action Button";

    const { getByText } = render(
      <ProviderWrapper>
        <Modal
          open={true}
          title="Test Modal"
          onClose={onCloseMock}
          maxWidth="sm"
          actions={<Button>{actionText}</Button>}
        >
          <div>Modal content goes here</div>
        </Modal>
      </ProviderWrapper>
    );

    expect(getByText(actionText)).toBeInTheDocument();
  });

  // Test for handling backdrop clicks
  it("closes on non-backdropClick reasons", () => {
    const onCloseMock = jest.fn();

    const { getByRole } = render(
      <ProviderWrapper>
        <Modal
          open={true}
          title="Test Modal"
          onClose={onCloseMock}
          maxWidth="sm"
        >
          <div>Modal content goes here</div>
        </Modal>
      </ProviderWrapper>
    );

    // Simulate the escape key down event
    fireEvent.keyDown(getByRole("dialog"), {
      key: "Escape",
      code: "Escape",
    });

    // Check if onClose was called
    expect(onCloseMock).toHaveBeenCalled();
  });
});
