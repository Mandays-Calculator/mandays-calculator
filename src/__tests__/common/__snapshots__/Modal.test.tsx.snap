// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Component renders and matches snapshot 1`] = `<DocumentFragment />`;
